src/ik.rs

```diff
@@ -212,42 +212,58 @@ where
         &self,
         arm: &SerialChain<T>,
         target_pose: &Isometry3<T>,
-        constraints: &Constraints,
+        operational_space: &[bool; 6],
+        ignored_joint_indices: &Vec<usize>,
     ) -> Result<DVector<T>, Error> {
-        let operational_space = define_operational_space(&constraints);
+        let required_dof = operational_space.iter().filter(|x| **x).count();
         let orig_positions = arm.joint_positions();
-        let dof = orig_positions.len();
+        let available_dof = arm.dof() - ignored_joint_indices.len();
+
         let t_n = arm.end_transform();
-        let err = calc_pose_diff_with_constraints(target_pose, &t_n, operational_space);
-        let orig_positions = arm.joint_positions();
+        let err = calc_pose_diff_with_constraints(target_pose, &t_n, *operational_space);
         let mut jacobi = jacobian(arm);
-        let use_dof = operational_space.iter().filter(|x| **x).count();
-        let mut removed_count = 0;
+        let mut num_removed_rows = 0;
         for (i, use_i) in operational_space.iter().enumerate() {
             if !use_i {
-                jacobi = jacobi.remove_row(i - removed_count);
-                removed_count += 1;
+                jacobi = jacobi.remove_row(i - num_removed_rows);
+                num_removed_rows += 1;
             }
         }
-        let positions_vec = if dof > use_dof {
+
+        for (i, joint_index) in ignored_joint_indices.iter().enumerate() {
+            jacobi = jacobi.remove_column(*joint_index - i);
+        }
+
+        let positions_vec = if available_dof > required_dof {
             const EPS: f64 = 0.0001;
             // redundant: pseudo inverse
             match self.nullspace_function {
                 Some(ref f) => {
                     let jacobi_inv = jacobi.clone().pseudo_inverse(na::convert(EPS)).unwrap();
-                    let d_q = jacobi_inv.clone() * err
-                        + (na::DMatrix::identity(dof, dof) - jacobi_inv * jacobi)
-                            * na::DVector::from_vec(f(&orig_positions));
+
+                    let mut subtask = na::DVector::from_vec(f(&orig_positions));
+                    for (i, joint_index) in ignored_joint_indices.iter().enumerate() {
+                        subtask = subtask.remove_row(*joint_index - i);
+                    }
+                    let mut d_q = jacobi_inv.clone() * err
+                        + (na::DMatrix::identity(available_dof, available_dof)
+                            - jacobi_inv * jacobi)
+                            * subtask;
+                    for joint_index in ignored_joint_indices.iter() {
```

Comment:
We don't need `iter()`

```suggestion
                    for joint_index in ignored_joint_indices {
```